In [119]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [120]:
df=pd.read_csv('relax_challenge/takehome_users.csv')

In [121]:
df

object_id        creation_time               name  \
0              1  2014-04-22 03:53:30     Clausen August   
1              2  2013-11-15 03:45:04      Poole Matthew   
2              3  2013-03-19 23:14:52  Bottrill Mitchell   
3              4  2013-05-21 08:09:28    Clausen Nicklas   
4              5  2013-01-17 10:14:20          Raw Grace   
...          ...                  ...                ...   
11995      11996  2013-09-06 06:14:15       Meier Sophia   
11996      11997  2013-01-10 18:28:37      Fisher Amelie   
11997      11998  2014-04-27 12:45:16        Haynes Jake   
11998      11999  2012-05-31 11:55:59       Faber Annett   
11999      12000  2014-01-26 08:57:12         Lima Tha�s   

                            email     creation_source  \
0        AugustCClausen@yahoo.com        GUEST_INVITE   
1          MatthewPoole@gustr.com          ORG_INVITE   
2      MitchellBottrill@gustr.com          ORG_INVITE   
3       NicklasSClausen@yahoo.com        GUEST_INVITE   
4              GraceRaw@yahoo.com        GUEST_INVITE   
...                           ...                 ...   
11995       SophiaMeier@gustr.com          ORG_INVITE   
11996      AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997         JakeHaynes@cuvox.de        GUEST_INVITE   
11998          mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999   ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
0                    1.398139e+09                         1   
1                    1.396238e+09                         0   
2                    1.363735e+09                         0   
3                    1.369210e+09                         0   
4                    1.358850e+09                         0   
...                           ...                       ...   
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
0                               0      11             10803.0  
1                               0       1               316.0  
2                               0      94              1525.0  
3                               0       1              5151.0  
4                               0     193              5240.0  
...                           ...     ...                 ...  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN  

[12000 rows x 10 columns]

Looking at the data set we are clear that it has 207917 rowas and 3 columns. Columns variables as time_stamp, 
user_id and number of times user has visited

In [122]:
df.shape

(12000, 10)

In [123]:
df1= pd.read_csv("relax_challenge/takehome_user_engagement.csv")

In [124]:
df1

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

df1 dataset has 207917 rows and 3 columns like time_stamp, user_id and number of times a user_id has visited

In [125]:
df.isnull().sum()

object_id                        0
creation_time                    0
name                             0
email                            0
creation_source                  0
last_session_creation_time    3177
opted_in_to_mailing_list         0
enabled_for_marketing_drip       0
org_id                           0
invited_by_user_id            5583
dtype: int64

In [126]:
df1.isnull().sum()

time_stamp    0
user_id       0
visited       0
dtype: int64

DATA CLEANING

To check if dataset 1 and dataset contains any null values. here Dataset df contains null values in last_session_creation_time
column where it has around 3177 null values and invited_by_user_id contains 5583 number of records as null

In [127]:
df.dropna(inplace=True)

In [128]:
df

object_id        creation_time               name  \
0              1  2014-04-22 03:53:30     Clausen August   
1              2  2013-11-15 03:45:04      Poole Matthew   
2              3  2013-03-19 23:14:52  Bottrill Mitchell   
3              4  2013-05-21 08:09:28    Clausen Nicklas   
4              5  2013-01-17 10:14:20          Raw Grace   
...          ...                  ...                ...   
11980      11981  2013-03-05 01:53:48          Fry Tyler   
11981      11982  2013-12-06 14:39:15      Barajas Maria   
11989      11990  2013-07-05 21:00:48          Juhl Aase   
11995      11996  2013-09-06 06:14:15       Meier Sophia   
11997      11998  2014-04-27 12:45:16        Haynes Jake   

                            email creation_source  last_session_creation_time  \
0        AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1          MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2      MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3       NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4              GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   
...                           ...             ...                         ...   
11980          TylerFry@gmail.com    GUEST_INVITE                1.364868e+09   
11981     MariaCBarajas@gmail.com      ORG_INVITE                1.386341e+09   
11989    AasePJuhl@jourrapide.com      ORG_INVITE                1.373144e+09   
11995       SophiaMeier@gustr.com      ORG_INVITE                1.378448e+09   
11997         JakeHaynes@cuvox.de    GUEST_INVITE                1.398603e+09   

       opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                             1                           0      11   
1                             0                           0       1   
2                             0                           0      94   
3                             0                           0       1   
4                             0                           0     193   
...                         ...                         ...     ...   
11980                         0                           0     110   
11981                         1                           1      57   
11989                         0                           0      25   
11995                         0                           0      89   
11997                         1                           1      83   

       invited_by_user_id  
0                 10803.0  
1                   316.0  
2                  1525.0  
3                  5151.0  
4                  5240.0  
...                   ...  
11980              5775.0  
11981              2527.0  
11989              3944.0  
11995              8263.0  
11997              8074.0  

[4776 rows x 10 columns]

In [129]:
df.isnull().sum()

object_id                     0
creation_time                 0
name                          0
email                         0
creation_source               0
last_session_creation_time    0
opted_in_to_mailing_list      0
enabled_for_marketing_drip    0
org_id                        0
invited_by_user_id            0
dtype: int64

it is better to remove the null values from a dataset to reduce baised decisions.
Now we can see none of the columns contain any null values 

In [130]:
df1

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

Converting series to DataFrame

In [132]:
#df1=pd.DataFrame(df1)
#df1

In [133]:
#df1.index=df1.time_stamp


In [134]:
df1

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [135]:
df1.drop(labels='time_stamp',axis=1,inplace=True)

In [136]:
df1

user_id  visited
0             1        1
1             2        1
2             2        1
3             2        1
4             2        1
...         ...      ...
207912    11996        1
207913    11997        1
207914    11998        1
207915    11999        1
207916    12000        1

[207917 rows x 2 columns]

In [137]:
df1 = pd.read_csv('relax_challenge/takehome_user_engagement.csv',encoding = "ISO-8859-1")
df1.time_stamp = pd.to_datetime(df1.time_stamp)
df1.index=df1.time_stamp
df1.drop(labels='time_stamp',axis=1,inplace=True)


In [138]:
df1

user_id  visited
time_stamp                           
2014-04-22 03:53:30        1        1
2013-11-15 03:45:04        2        1
2013-11-29 03:45:04        2        1
2013-12-09 03:45:04        2        1
2013-12-25 03:45:04        2        1
...                      ...      ...
2013-09-06 06:14:15    11996        1
2013-01-15 18:28:37    11997        1
2014-04-27 12:45:16    11998        1
2012-06-02 11:55:59    11999        1
2014-01-26 08:57:12    12000        1

[207917 rows x 2 columns]

In [139]:
df_sum = df1.groupby([pd.Grouper(freq='W'),'user_id']).sum()

In [140]:
df_sum

visited
time_stamp user_id         
2012-06-03 563            1
           1693           1
           1995           1
           2120           1
           2136           1
...                     ...
2014-06-08 11869          1
           11885          1
           11895          3
           11906          1
           11924          3

[68807 rows x 1 columns]

Now to find out a user_id who has logged into the product on three separate
days in at least one sevenday
period , identify which factors predict future user adoption



In [141]:
df_adopted = df_sum[df_sum.visited>=3].unstack(level=1).melt()
adopted_users = pd.DataFrame(df_adopted.user_id.unique(),index=range(df_adopted.user_id.unique().shape[0]),columns=['user_id'])

In [142]:
df_adopted


NaN  user_id  value
0       visited     1693    3.0
1       visited     1693    NaN
2       visited     1693    3.0
3       visited     1693    NaN
4       visited     1693    3.0
...         ...      ...    ...
151720  visited     8460    NaN
151721  visited     8460    NaN
151722  visited     8460    NaN
151723  visited     8460    NaN
151724  visited     8460    3.0

[151725 rows x 3 columns]

In [143]:
adopted_users

user_id
0        1693
1         728
2       11764
3        5297
4        6171
...       ...
1440     7868
1441     7927
1442     9870
1443    10746
1444     8460

[1445 rows x 1 columns]

In [144]:
df_join = df.merge(adopted_users,how='inner',left_on='object_id',right_on='user_id')
df_join.shape

(817, 11)

Here working on the two datasets where dataset(df) is user table ( "takehome_users" ) with data on 12,000 users who signed up for the
product in the last two years.

and another dataset (df1) is usage summary table ( "takehome_user_engagement" ) that has a row for each day
that a user logged into the product

While analysing both the datasets we have removed all the null values and columns which is not making sense and
will not be able to provide any valuable insights about the dataset.

AFter grouping it as if any user_id has visited the site more than or equal to 3 times, it is considered to as
adopted user. Breaking down the data to that extend we have found the user_ids which are prone to become 
adopted. then merging it with first dataset (df) is making sense where we can get all the data of the user_ids
along with email, how the user was invited , org_id etc.